test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/CacheTagHelperTest.cs

```diff
@@ -777,6 +782,71 @@ public void UpdateCacheEntryOptions_CopiesTriggersFromEntryLink()
             Assert.Null(cachedValue);
         }
 
+        [Fact]
+        public async Task ProcessAsync_ComputesValueOnce_WithConcurrentRequests()
+        {
+            // Arrange
+            var id = "unique-id";
+            var childContent = "some-content";
+            var resetEvent1 = new ManualResetEvent(false);
+            var resetEvent2 = new ManualResetEvent(false);
+            var calls = 0;
+            var cache = new MemoryCache(new MemoryCacheOptions());
+
+            TagHelperContext tagHelperContext1 = GetTagHelperContext(id + 1);
+            TagHelperContext tagHelperContext2 = GetTagHelperContext(id + 2);
+
+            TagHelperOutput tagHelperOutput = new TagHelperOutput(
+                "cache",
+                new TagHelperAttributeList(),
+                getChildContentAsync: (useCachedResult, encoder) =>
+                {
+                    calls++;
+                    resetEvent2.Set();
+
+                    var tagHelperContent = new DefaultTagHelperContent();
+                    tagHelperContent.SetHtmlContent(childContent);
+                    return Task.FromResult<TagHelperContent>(tagHelperContent);
+                });
+
+            var cacheTagHelper1 = new CacheTagHelper(cache, new HtmlTestEncoder())
+            {
+                ViewContext = GetViewContext(),
+                Enabled = true
+            };
+
+            var cacheTagHelper2 = new CacheTagHelper(cache, new HtmlTestEncoder())
+            {
+                ViewContext = GetViewContext(),
+                Enabled = true
+            };
+
+            // Act
+
+            var task1 = Task.Run(async () =>
+            {
+                resetEvent1.WaitOne();
```

Comment:
Can you add the wait duration here so the test doesn't hang forever in case something goes wrong? ~5-10 seconds should be good.